src/sync.cpp

```diff
@@ -331,6 +323,85 @@ void assignFilesystemIdsImpl(bool& success, Sync& sync, MegaApp& app, handleloca
     }
 }
 
+// Assigns fs IDs from `files` to those `localnodes` that match the fingerprints found in `files`.
+// If there are multiple matches we apply a best-path heuristic.
+size_t assignFilesystemIdsImpl(const FingerprintCache& fingerprints, FingerprintLocalNodeMap& localnodes,
+                               FingerprintFileMap& files, handlelocalnode_map& fsidnodes, const string& localseparator)
+{
+    string nodePath;
+    string accumulated;
+    size_t assignmentCount = 0;
+    for (const auto& fp : fingerprints.all())
+    {
+        const auto nodeRange = localnodes.equal_range(&fp);
+        const auto nodeCount = std::distance(nodeRange.first, nodeRange.second);
+        if (nodeCount <= 0)
+        {
+            continue;
+        }
+
+        const auto fileRange = files.equal_range(&fp);
+        const auto fileCount = std::distance(fileRange.first, fileRange.second);
+        if (fileCount <= 0)
+        {
+            // without files we cannot assign fs IDs to these localnodes, so no need to keep them
+            localnodes.erase(nodeRange.first, nodeRange.second);
+            continue;
+        }
+
+        struct Element
+        {
+            int score;
+            handle fsid;
+            LocalNode* l;
+        };
+        std::vector<Element> elements;
+        elements.reserve(nodeCount * fileCount);
+
+        for (auto nodeIt = nodeRange.first; nodeIt != nodeRange.second; ++nodeIt)
+        {
+            auto l = nodeIt->second;
+            if (l != &l->sync->localroot) // never assign fs ID to the root localnode
+            {
+                nodePath.clear();
+                l->getlocalpath(&nodePath, false, &localseparator);
+                for (auto fileIt = fileRange.first; fileIt != fileRange.second; ++fileIt)
+                {
+                    const auto& filePath = fileIt->second.path;
+                    const auto score = computeReversePathMatchScore(accumulated, nodePath, filePath, localseparator);
+                    if (score > 0) // leaf name must match
```

Comment:
why? don't we want a match of a renamed file? what am I missing?

Reply:
Thanks for the the explanation Christian. As a matter of fact I was thinking that would be the reason later on, after posting the comment. We might want to add a word at the definition of `assignFilesystemIds`, or at `assignFilesystemIdsImpl` or `computeReversePathMatchScore` explaining that _renames_ are considered misses due to the heuristics computation  (and that `if` prevents them from being assigned). Anyway: no showstopper